### PR TITLE
fix(security): delete grant once used

### PR DESCRIPTION
**Pull request by mdtro**
When an `ApiToken` is created from an `ApiGrant` we should delete that specific grant from the database in order prevent reuse of the same code for token exchanges.

Note, we already do this in the "grant exchanger". This particular code path does not use the generic `GrantExchanger`. I'm not sure why, but that's something we can come back to refactor. 
https://github.com/getsentry/sentry/blob/80f9de5890cac1a7d70252f7f4818baaf2155dd9/src/sentry/mediators/token_exchange/grant_exchanger.py#L37-L39

### CODE REVIEW

File: src/sentry/models/apitoken.py [hunk 1]
```diff
@@ -83,12 +84,18 @@ def handle_async_replication(self, region_name: str, shard_identifier: int) -> N
...
+    def from_grant(cls, grant: ApiGrant):
         with transaction.atomic(router.db_for_write(cls)):
-            return cls.objects.create(
+            api_token = cls.objects.create(
                 application=grant.application, user=grant.user, scope_list=grant.get_scopes()
             )
 
+            # remove the ApiGrant from the database to prevent reuse of the same
+            # authorization code
+            grant.delete()
```
Comment (ykamo001) [inline src/sentry/models/apitoken.py:95]: do we want to delete this forever, or do we want to soft delete it so we have some sort of trail/audit for at least a few days to help debug and support customers?

Comment (ykamo001) [reply src/sentry/models/apitoken.py:95]: there's a pipeline that deletes objects/rows after 30 days that we can just use to help clean up

Comment (mdtro) [reply src/sentry/models/apitoken.py:95]: The `ApiGrant` needs to be unusable as soon as it is exchanged for a token. We can soft delete these, but I'd like to merge this as it resolves an immediate security bug. I can issue new PRs for the necessary model changes to support soft deletions.

I'll take a look at the cleanup script. It should be removing these that are expired regardless of how old they actually are. Ironically, the cleanup script purging these might have been created to deal with the fallout of this bug. 🤔

Comment (ykamo001) [reply src/sentry/models/apitoken.py:95]: Yeah, sure thing

Comment (mdtro) [reply src/sentry/models/apitoken.py:95]: Ah it does delete them based on the expiring time. 😄 

https://github.com/getsentry/sentry/blob/485f3a42f9a977d4e23b93b4580ad2837ba1027f/src/sentry/runner/commands/cleanup.py#L251-L268

**Review by ykamo001 — APPROVED** [review]

**Pull request merged**